sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -244,6 +244,7 @@ class Analyzer(
       ExtractGenerator ::
       ResolveGenerate ::
       ResolveFunctions ::
+      new SubstituteUnresolvedOrdinals(conf) ::
```

Comment:
Why we need to add this rule in the resolution batch?

Reply:
> Why we need to add this rule in the resolution batch?

Since `cube/rollup` can be a function.

Reply:
Still not sure about why we cannot transform `cube(1, 2)` -> `cube(unresolvedordinal(1), unresolvedordinal(2))` in the substitution batch, then resolve them in the resolution batch. Any technical difficulty?

Reply:
> Still not sure about why we cannot transform `cube(1, 2)` -> `cube(unresolvedordinal(1), unresolvedordinal(2))` in the substitution batch, then resolve them in the resolution batch. Any technical difficulty?

since when we execute substitution batch, `cube(1,2)` still is UnresolvedFunction('cube(1,2)').  We need to solve it to `Cube(unresolvedordinal(1), unresolvedordinal(2))` after `ResolveFunction`

Reply:
Ah, I got it. Thanks. But, it looks a bit weird for the resolution batch to have the substitution rule. Instead, we cannot support CUBE(...) and ROLLUP(...) in the parser level so as to avoid `UnresolvedFunction('cube(1,2)')`? cc: @cloud-fan

Reply:
> Ah, I got it. Thanks. But, it looks a bit weird for the resolution batch to have the substitution rule. Instead, we cannot support CUBE(...) and ROLLUP(...) in the parser level so as to avoid `UnresolvedFunction('cube(1,2)')`? cc: @cloud-fan

These days. I am thinking about this too. Since as you mentioned in https://github.com/apache/spark/pull/30144#issuecomment-716353950,  `grouping sets` can be a expression like `cube` , `rollup`.

Current parser level can't support it smoothly. Maybe we can reconstruct 
```

aggregationClause
    : GROUP BY groupingExpressions+=expression (',' groupingExpressions+=expression)* (
      WITH kind=ROLLUP
    | WITH kind=CUBE
    | kind=GROUPING SETS '(' groupingSet (',' groupingSet)* ')')?
    | GROUP BY kind=GROUPING SETS '(' groupingSet (',' groupingSet)* ')'
    ;

groupingSet
    : '(' (expression (',' expression)*)? ')'
    | expression
    ;

```

Reply:
okay, thanks for the check. Could you try it in a separate PR first?

Reply:
> okay, thanks for the check. Could you try it in a separate PR first?

Working on this,

Reply:
FYI @cloud-fan  https://github.com/apache/spark/pull/30199#discussion_r515042507